### ISSUE TITLE
OAuth 2.0 blocked by API key credentials

### ISSUE THREAD

**Issue by AlanPew** [Esri/arcgis-runtime-samples-ios#1101]
Hi, I am currently using Oauth 2.0 to access a portal id that is not associated with my developer account using AGSPortal . Had trouble accessing this until I commented out the api license key authorization in the app delegate. Can now access the portalId and it displays on the app, but now shows "FOR DEVELOPER USE ONLY" because the api key was not authorized. The portal is accessed via AGSPortal using an instance of AGSCredential. 

 Is this an expected behavior of the ArcGIS iOS SDK or are the API key credentials superseding (and blocking) the OAuth 2.0 credential authorization because the portalID i'm trying to access is not associated with my developer account?  

Thanks!

PS. The ArcGIS ios runtime samples are wonderful! Thank you for making this code available.

**Comment by yo1995** [comment]
Hi Alan,

I'd suggest you to re-post this question on GeoNet: https://community.esri.com/t5/arcgis-runtime-sdk-for-ios-questions/bd-p/arcgis-runtime-sdk-for-ios-questions , so that our other developers may provide more insight. 🙂 

In most cases, the API key does not conflict with OAuth 2.0, but it's recommended not to mix them up in your app. Read more about the security methods here: https://developers.arcgis.com/documentation/mapping-apis-and-services/security/security-best-practices/

- In my personal experience, API keys are more intended for the read-only access in public-facing app, e.g. loading a map, solving a route, doing geocoding, etc.
- While the OAuth is intended for access a group of resources under an account/organization. OAuth also provides most of the capabilities that an API key does (by consuming credits), if the account has access to the content and services.

---

In terms of removing the watermark, a "license string". It is a different thing than the authentication. Read more at https://developers.arcgis.com/documentation/mapping-apis-and-services/deployment/distribution/#native-applications 

There does exist some ordering for the API key and OAuth credential (?). I'll invite some expert eyes to take a look.

cc @dg0yal

**Comment by yo1995** [comment]
Further info from our team:

If you set a global API key, then all [`AGSAPIKeyResource`](https://developers.arcgis.com/ios/api-reference/protocol_a_g_s_a_p_i_key_resource-p.html) objects will use it and won’t challenge for credentials. 

Therefore, if your portal item is one of these classes, try to avoid using global API key, but only set API key on those resources that have nothing to do with your ArcGIS account. So that the portal item can correctly prompt for OAuth challenge.

**Comment by AlanPew** [comment]
Thank you so much for your detailed and quick response @yo1995!

When I uncommented out the application.license() call in the app delegate, it was calling      AGSArcGISRuntimeEnvironment.apiKey = .apiKey and getting this error message:

The API key '*xxxxxx' is invalid, expired, or does not have access to the resource:
- 'https://xxxxxxxxxxx/sharing/rest/content/items/xxxxxxxxxx'. 
Please visit https://developers.arcgis.com/api-keys to configure the API key to access the resource, or to create a new API key.
 in addition to the setting the license key:
 (try AGSArcGISRuntimeEnvironment.setLicenseKey(.licenseKey)

 I am fairly new to the ArcGIS iOS SDK and didn't realize both the license key and api were being set in the application.license() function call. I was commenting out the application.license() call to avoid setting the api key, which meant the license key wasn't being set and the watermark appeared. When I commented out AGSArcGISRuntimeEnvironment.apiKey = .apiKey in  application,license(), I was able to set the license key properly and my OAuth 2.0 credentials did not conflict with the apiKey. No longer seeing the "Licensed for Developer Use Only" watermark.

Thank you so much for your help!!! Looking forward to developing with this SDK! Closing this issue now.

**Comment by yo1995** [comment]
Glad to know that you've figured out the licensing options!

> When I uncommented out the application.license() call in the app delegate...

Sorry for this confusion 😅  it was a makeshift that I used in the last release to facilitate our daily development. With changes in #1092 and the [new release version 100.12.0.1](https://github.com/Esri/arcgis-runtime-samples-ios/releases/tag/v100.12.0.1) it shouldn't be a problem anymore!

You are welcome to open another issue or ask on [GeoNet](https://community.esri.com/t5/arcgis-runtime-sdk-for-ios-questions/bd-p/arcgis-runtime-sdk-for-ios-questions) in the future. Happy coding with Runtime SDK! 🥰
